src/kv/snapshot.h

```diff
@@ -38,9 +63,33 @@ namespace kv
       return version;
     }
 
-    std::vector<uint8_t> serialise(
-      const std::shared_ptr<AbstractTxEncryptor>& encryptor) override
+    size_t serialised_size(
+      const std::shared_ptr<AbstractTxEncryptor>& encryptor) const override
+    {
+      KvStoreSizeCounterSerialiser serialiser(
+        encryptor,
+        {0, version},
+        kv::EntryType::Snapshot,
+        0,
+        {},
+        ccf::no_claims(),
+        true /* historical_hint */);
+
+      serialise(serialiser);
+
+      return serialiser.get_serialised_size();
+    }
+
+    void serialise(
+      const std::shared_ptr<AbstractTxEncryptor>& encryptor,
+      std::span<uint8_t> serialised_snapshot) override
     {
+      CCF_ASSERT_FMT(
+        serialised_snapshot.size() == serialised_size(encryptor),
```

Comment:
As discussed, passing a single size through all the serialize() calls should allow us to do this check very cheaply and without recomputation: they all bound check, and we should end on 0.